app/services/data_formatter.rb

```diff
@@ -0,0 +1,57 @@
+class DataFormatter
+  def initialize(purchases)
+    @purchases = purchases
+  end
+
+  def category_totals
+    Category.all.map do |c|
+      [c.name, sum_purchases_in_category(c)]
+    end
+  end
+
+  def category_totals_by_month
+    first_purchase = Purchase.all.order(:date).first[:date]
+    last_purchase = Purchase.all.order(:date).last[:date] # this could be Date.today rather than most recent purchase
+    months_in_range = (first_purchase..last_purchase).map { |d| d.strftime('%B %Y') }.uniq
+
+    sum_purchases_by_month(purchases_by_category, months_in_range)
+  end
+
+  private
+
+  def sum_purchases_in_category(category)
+    @purchases.where(subcategory: category.subcategories).to_a.map(&:amount).sum
+  end
+
+  def purchases_by_category
+    totals = Category.all.map do |c|
+      extract_purchases_by_category(c).map { |p| { name: c.name, date: p[:date], amount: p[:amount] } }
+    end
+
+    totals.flatten.group_by { |n| n[:name] }
+  end
+
+  def extract_purchases_by_category(category)
+    @purchases.where(subcategory: category.subcategories).map { |p| { date: p.date, amount: p.amount } }
+  end
+
+  def sum_purchases_by_month(purchases_by_category, months_in_range)
+    summaries = purchases_by_category.map do |key, value|
+      { name: key, data: value.group_by_month { |n| n[:date] }.map do |key, value|
+        [key.strftime('%B %Y'), value.map do |v|
+          v[:amount]
+        end.sum]
+      end }
+    end
+
+    summaries.each do |c|
```

Comment:
Also, I think I'd extract this block into a new method, e.g. `ensure_all_months_present`.
`first_purchase`, `last_purchase`, and `months_in_range` can actually just be defined in there, rather than passed down from `category_totals_by_month`.

Reply:
I was playing around with this, here's the implementation I came up with:

```ruby
  def ensure_all_months_present(summaries)
    first_purchase = Purchase.all.order(:date).first[:date] # TODO: better query
    last_purchase = Purchase.all.order(:date).last[:date]
    months_in_range = (first_purchase..last_purchase).map { |d| d.strftime('%B %Y') }.uniq

    summaries.each do |summary|
      month_totals = months_in_range.map do |month|
        existing_total = summary[:data].detect { |s| s.first == month }

        existing_total ? existing_total : [month, 0]
      end

      summary[:data] = month_totals
    end
  end
```
Not sure if it is the 'best' way to tackle this, but possibly food for thought.

Reply:
Oh sorry, I didn't mean that you had to do it that way! It was just what I came up with.